### PR TITLE
bpo-33012: Fix invalid function cast warnings with gcc 8 for METH_NOARGS

### PR DESCRIPTION
**Pull request by siddhesh**
gcc 8 has added a new warning heuristic to detect invalid function casts and a stock python build seems to hit that warning quite often.  The most common is the cast of a METH_NOARGS function (that uses just one argument) to a PyCFunction.  Fix this by adding a dummy argument to all the affected callback functions.

I have split the changes by directory and also made a separate commit where the change was not just a trivial dummy argument addition, as is the case in a4ee3ad and 58eaf31.

https://bugs.python.org/issue33012

### CODE REVIEW

File: Doc/extending/newtypes.rst [hunk 1]
```diff
@@ -464,8 +464,9 @@ The method is implemented as a C function that takes a :class:`Noddy` (or
 :class:`Noddy` subclass) instance as the first argument.  Methods always take an
 instance as the first argument. Methods often take positional and keyword
 arguments as well, but in this case we don't take any and don't need to accept
-a positional argument tuple or keyword argument dictionary. This method is
-equivalent to the Python method::
+a positional argument tuple or keyword argument dictionary. We add an unused
+argument to allow for a safe cast to a PyCFunction. This method is equivalent to
```
Comment (serhiy-storchaka) [inline Doc/extending/newtypes.rst]:
```
:c:type:`PyCFunction`
```

File: Objects/odictobject.c [hunk 2]
```diff
@@ -1372,8 +1378,8 @@ static PyMethodDef odict_methods[] = {
 
     /* explicitly defined so we can align docstrings with
      * collections.OrderedDict */
-    {"__delitem__",     (PyCFunction)odict_mp_ass_sub,  METH_NOARGS,
-     odict_delitem__doc__},
+    {"__delitem__",     (PyCFunction)odict_mp_ass_sub,
+     METH_VARARGS | METH_KEYWORDS, odict_delitem__doc__},
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: This looks incorrect. But an existing code looks incorrect too. I suggest to revert this change (and all other questionable changes) and left them to a separate issue -- [bpo-33031](https://bugs.python.org/issue33031).

File: Objects/odictobject.c [hunk 3]
```diff
@@ -1466,7 +1472,7 @@ odict_dealloc(PyODictObject *self)
 /* tp_repr */
 
 static PyObject *
-odict_repr(PyODictObject *self)
+odict_repr(PyODictObject *self, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: It should be compatible with `reprfunc`, not `PyCFunction`.

File: Objects/odictobject.c [hunk 4]
```diff
@@ -1609,15 +1615,15 @@ odict_richcompare(PyObject *v, PyObject *w, int op)
 /* tp_iter */
 
 static PyObject *
-odict_iter(PyODictObject *od)
+odict_iter(PyODictObject *od, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: It should be compatible with `getiterfunc`, not `PyCFunction`.

File: Modules/_collectionsmodule.c [hunk 5]
```diff
@@ -1532,7 +1532,7 @@ deque_bool(dequeobject *deque)
 }
 
 static PyObject *
-deque_get_maxlen(dequeobject *deque)
+deque_get_maxlen(dequeobject *deque, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Modules/_collectionsmodule.c]: It is `getter`.
```
void *Py_UNUSED(ignored)
``

File: Objects/descrobject.c [hunk 6]
```diff
@@ -439,7 +439,7 @@ calculate_qualname(PyDescrObject *descr)
 }
 
 static PyObject *
-descr_get_qualname(PyDescrObject *descr)
+descr_get_qualname(PyDescrObject *descr, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Objects/descrobject.c]:
```
void *Py_UNUSED(ignored)
```

File: Objects/descrobject.c [hunk 7]
```diff
@@ -1139,7 +1139,7 @@ wrapper_text_signature(wrapperobject *wp, void *closure)
 static PyObject *
 wrapper_qualname(wrapperobject *wp)
```
Comment (serhiy-storchaka) [inline Objects/descrobject.c]: Needs the second argument of type `void *`. As well as other getters.

File: Objects/longobject.c [hunk 8]
```diff
@@ -4554,7 +4554,7 @@ long_or(PyObject *a, PyObject *b)
 }
 
 static PyObject *
-long_long(PyObject *v)
+long_long(PyObject *v, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Objects/longobject.c:4557]: It is used both as `PyCFunction` and `getter`. It would be safer to declare the second parameter as `void *`.

File: Objects/odictobject.c [hunk 9]
```diff
@@ -887,14 +887,20 @@ PyDoc_STRVAR(odict_init__doc__,
 \n\
         ");
 
+static int odict_tp_init(PyObject *, PyObject *, PyObject *);
 /* forward */
-static int odict_init(PyObject *self, PyObject *args, PyObject *kwds);
+static PyObject *
+odict_init(PyObject *self, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Objects/odictobject.c]: I suggest to left this for a separate issue.

File: Modules/_collectionsmodule.c [hunk 10]
```diff
@@ -512,7 +512,7 @@ deque_inplace_concat(dequeobject *deque, PyObject *other)
 }
 
 static PyObject *
-deque_copy(PyObject *deque)
+deque_copy(PyObject *deque, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline Modules/_collectionsmodule.c:515]: Explicit cast to `PyCFunction` at line 1590 is not needed.

The same with other functions which exactly match the signature of `PyCFunction`.

File: Objects/descrobject.c (hunk 7 above)
Comment (siddhesh) [reply Objects/descrobject.c]: I've made a separate bug report for the getters: issue33029.

File: Objects/odictobject.c (hunk 9 above)
Comment (siddhesh) [reply Objects/odictobject.c]: OK, I'll drop this and the other patch from this PR and make a different one.

File: Modules/_collectionsmodule.c (hunk 10 above)
Comment (siddhesh) [reply Modules/_collectionsmodule.c:515]: I've kept the casts in place because we will eventually need it.  To fully fix the warnings, the `ml_meth` type should be like so:

`PyObject *(*) (PyObject *, PyObject *, ...)`

so that it matches with all of the allowed method types provided that there is an explicit cast in place.  Right now I am not sure if I should change the `PyCFunction` type to that or to make a new `PyCFunctionAny` just for `ml_meth`.  Since we're on this topic, what do you think is the preferred change?  I suppose both change API, but not in a completely incompatible way.

File: Modules/_collectionsmodule.c (hunk 5 above)
Comment (siddhesh) [reply Modules/_collectionsmodule.c]: I'm tackling getters in a separate patch, so I'll drop this.

File: Objects/descrobject.c (hunk 6 above)
Comment (siddhesh) [reply Objects/descrobject.c]: Likewise, I've got getter changes in another branch, so I'll fix it there.

File: Objects/odictobject.c (hunk 3 above)
Comment (siddhesh) [reply Objects/odictobject.c]: I'll drop this one too since it looks like it will need an additional wrapper like odict_init.

File: Objects/odictobject.c (hunk 4 above)
Comment (siddhesh) [reply Objects/odictobject.c]: Likewise, I'll deal with it separately.

File: Objects/longobject.c (hunk 8 above)
Comment (siddhesh) [reply Objects/longobject.c:4557]: Done.

File: Objects/odictobject.c (hunk 2 above)
Comment (siddhesh) [reply Objects/odictobject.c]: OK, dropped from this PR.

**Review by siddhesh — COMMENTED** [review]
Thanks for the review, I'm pushing a version with most of the fixes you suggested, except for removal of the cast. I've kept the casts in place because we will eventually need it. To fully fix the warnings, the `ml_meth` type should be like so:

`PyObject *(*) (PyObject *, PyObject *, ...)`

so that it matches with all of the allowed method types provided that there is an explicit cast in place. Right now I am not sure if I should change the `PyCFunction` type to that or to make a new `PyCFunctionAny` just for `ml_meth`. Since we're on this topic, what do you think is the preferred change? I suppose both change API, but not in a completely incompatible way.

File: PC/bdist_wininst/install.c [hunk 11]
```diff
@@ -631,7 +631,7 @@ static PyObject *PyMessageBox(PyObject *self, PyObject *args)
     return g_Py_BuildValue("i", rc);
 }
 
-static PyObject *GetRootHKey(PyObject *self)
+static PyObject *GetRootHKey(PyObject *self, PyObject *Py_UNUSED(ignored))
```
Comment (serhiy-storchaka) [inline PC/bdist_wininst/install.c:634]: Casting to `PyCFunction` below no longer needed. And this is the only use of `PyCFunction`, thus its declaration can be removed.

**Comment by siddhesh** [comment]
Hi, here's the new patchset with dropped casts and resolved conflicts.

File: Modules/_collectionsmodule.c (hunk 10 above)
Comment (serhiy-storchaka) [reply Modules/_collectionsmodule.c:515]: The casts are not needed if the signature matches `PyCFunction`. It is still needed if the function has type `PyCFunctionWithKeywords`, `PyNoArgsFunction`, etc. Removing unneeded casts will clean up the code and can prevent introducing new bugs when the signature be changed unintentionally.

File: PC/bdist_wininst/install.c (hunk 11 above)
Comment (siddhesh) [reply PC/bdist_wininst/install.c:634]: Ah I missed this one, I'll fix it.

**Review by serhiy-storchaka — APPROVED** [review]
